doc/md/versioned: upgrading to versioned migrations (#3095)
ent/ent: doc/md/versioned/07-verifying-safety.md

---
title: Verifying migration safety
id: verifying-safety
---
## Supporting repository

The change described in this section can be found in
[PR #8](https://github.com/rotemtam/ent-versioned-migrations-demo/pull/8/files)
in the supporting repository.

## Verifying migration safety

As the database is a critical component of our application, we want to make sure that when we 
make changes to it, we don't break anything. Ill-planned migrations can cause data loss, application
downtime and other issues.  Atlas provides a mechanism to verify that a migration is safe to run.
This mechanism is called [migration linting](https://atlasgo.io/versioned/lint) and in this section
we will show how to use it to verify that our migration is safe to run.

## Linting the migration directory

To lint our migration directory we can use the `atlas migrate lint` command.
To demonstrate this, let's add a migration that drops the `blogs` table:

```shell
atlas migrate new drop_blogs --dir file://ent/migrate/migrations
```
Edit the newly created migration file and add the following SQL statement:

```sql

> drop table blogs;

Reviewer: data-dependent changes are more attractive.

Author: replaced

```text
20221115124032_drop_blogs.sql: destructive changes detected:

        L1: Dropping table "blogs"

```
Atlas detected that the migration is unsafe to run and prevented us from running it.
In this case, Atlas classified this change as a destructive change. Destructive changes are
changes that cause data loss and are not reversible. In this case, the `blogs` table is dropped,
while the table may be recreated, it will not be restored with the original data.

Atlas can detect many more types of issues, for a full list, see the [Atlas documentation](https://atlasgo.io/lint/analyzers).

## Linting our migration directory in CI

In the previous section, we saw how to lint our migration directory locally. In this section,
we will see how to lint our migration directory in CI. This way, we can make sure that our migration
history is safe to run before we merge it to the main branch.

[GitHub Actions](https://github.com/features/actions) is a popular CI/CD
product from GitHub. With GitHub Actions, users can easily define workflows
that are triggered in various lifecycle events related to a Git repository.
For example, many teams configure GitHub actions to run all unit tests in

> a repository on each change that is committed to a repository.

Reviewer 2: ```suggestion
a repository on each change that is committed to its mainline branch.
```

One of the powerful features of GitHub Actions is its extensibility: it is
very easy to package a piece of functionality as a module (called an "action")

> that can later be re-used by many projects.

Reviewer 2: ```suggestion
that can later be reused by many projects.
```

Teams using GitHub that wish to ensure all changes to their database schema are safe
can use the [`atlas-action`](https://github.com/ariga/atlas-action) GitHub Action.

This action is used for [linting migration directories](/versioned/lint)
using the `atlas migrate lint` command. This command  validates and analyzes the contents
of migration directories and generates insights and diagnostics on the selected changes:

> * Ensure the migration history can be replayed from any point at time.

Reviewer 2: ```suggestion
* Ensure the migration history can be replayed from any point in time.
```

          MYSQL_DATABASE: dev
        ports:
          - "3306:3306"
        options: >-
          --health-cmd "mysqladmin ping -ppass"
          --health-interval 10s
          --health-start-period 10s
          --health-timeout 5s
          --health-retries 10
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3.0.1
        with:
          fetch-depth: 0 # Mandatory unless "latest" is set below.
      - uses: ariga/atlas-action@v0
        with:
          dir: ent/migrate/migrations
          dev-url: mysql://root:pass@localhost:3306/dev
```
Now, whenever we make a pull request with a potentially unsafe migration, the Atlas
GitHub action will run and report the linting results. For example, for our destructive change:
![](https://atlasgo.io/uploads/images/atlas-ci-report-comment.png)

For more in depth documentation, see the [atlas-action](https://atlasgo.io/integrations/github-actions)
docs on the Atlas website.

Let's fix the issue by removing our dangerous migration and re-hashing the
migration directory:
```text
rm ent/migrate/migrations/20221115124032_drop_blogs.sql
atlas migrate hash --dir file://ent/migrate/migrations
```

> Re-running `atlas migrate lint` we can see that the migration directory doesn't

Reviewer 2: ```suggestion
Re-running `atlas migrate lint`, we can see that the migration directory doesn't
```

contain any unsafe changes:

```text
atlas migrate lint --dev-url mysql://root:pass@localhost:3306/dev --dir file://ent/migrate/migrations --latest 1
```

Because no issues are found, the command will exit with a zero exit code and no output. 

## Conclusion

In this section, we saw how to use Atlas to verify that our migration is safe to run both
locally and in CI.

This wraps up our tutorial on how to upgrade your Ent project from
automatic migration to versioned migrations. To recap, we learned how to:

* Enable the versioned migrations feature-flag
* Create a script to automatically plan migrations based on our desired Ent schema
* Upgrade our production database to use versioned migrations with Atlas
* Plan custom migrations for our project

> * Verifying migrations are safe using `atlas migrate lint`

Reviewer 2: ```suggestion
* Verify migrations safely using `atlas migrate lint`
```
